[package] Make exporters write to buffer in fbcode

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54303 [package] Make exporters write to buffer in fbcode**

**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

Differential Revision: [D27180839](https://our.internmc.facebook.com/intern/diff/D27180839)

Commit:
[package] Make exporters write to buffer in fbcode

**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

[ghstack-poisoned]

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -615,6 +615,11 @@ def close(self):
         if self.buffer:
             self.buffer.flush()
 
+        # If self.buffer is a BytesIO, rewind it in anticipation
+        # of a read in the near future.
+        if isinstance(self.buffer, io.BytesIO):
+            self.buffer.seek(0)
```
I don't know if this is really that great of an idea, but it does help obviate the need to add `buffer.seek(0)` all over the place in unit tests.

suo replied:
Probably not a good idea.
- Convention of Python standard library is not to do this (in particular `pickle` does not do this)
- We shouldn't assume ownership over the io stream.

suo commented on test/package/common.py:
```diff
@@ -11,6 +12,10 @@ def __init__(self, *args, **kwargs):
         self._temporary_files = []
 
     def temp(self):
+        # Use a buffer instead of a temporary file in fbcode.
```
I'm not sure if we want to have the behavior be different between fbcode/external. The benefit is that we get both paths tested well, the disadvantage is that all test code needs to handle both cases, and it makes the test logic slightly more confusing (not obvious that behavior is different when debugging an individual test).

What about: use `BytesIO` everywhere as convention, but have the tests in `test_save_load`  explicitly test the file-based path and disable those tests in sandcastle? It's more work, but I think it's good to have the tests be very explicit. Could be convinced either way though.

SplitInfinity replied:
I made this change and pushed again. The only thing I don't like about it is that people writing tests have to remember to `buffer.seek(0)` all the time.

What about introducing a context manager that automatically rewinds the test buffer that its `__enter__` returns:
```
with PackageTestBuffer() as buffer:
    # use buffer in exporter

# buffer is automatically rewound
# use buffer in importer
```

It is not intuitive that buffer can be used outside of the with statement but Python doesn't have scope anyway...

suo replied:
Dunno if that's much better, you have to remember to do something special either way. At least `io` is a standard thing.

Commit:
Update on "[package] Make exporters write to buffer in fbcode"


**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

Differential Revision: [D27180839](https://our.internmc.facebook.com/intern/diff/D27180839)

[ghstack-poisoned]

Commit:
Update on "[package] Make exporters write to buffer in fbcode"


**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

Differential Revision: [D27180839](https://our.internmc.facebook.com/intern/diff/D27180839)

[ghstack-poisoned]

Commit:
Update on "[package] Make exporters write to buffer in fbcode"


**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

Differential Revision: [D27180839](https://our.internmc.facebook.com/intern/diff/D27180839)

[ghstack-poisoned]

Commit:
Update on "[package] Make exporters write to buffer in fbcode"


**Summary**
Creating temporary files can cause problem in fbcode. This commit
updates the packaging tests so that exporters write to a memory
buffer when tests run in fbcode.

**Test Plan**
Continuous integration.

Differential Revision: [D27180839](https://our.internmc.facebook.com/intern/diff/D27180839)

[ghstack-poisoned]
